Add JDBC record reader (#362)

dalsh opened the pull request:
## What changes were proposed in this pull request?

This adds the JDBC record reader mentioned in #362 . Work is still in progress but I'm putting it here if you want to start reviewing it anyway.

## How was this patch tested?

This should be fully covered in JDBCRecordReaderTest

agibsonccc commented on datavec-jdbc/src/main/java/org/datavec/api/records/reader/impl/jdbc/JDBCRecordReader.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    @Override
+    public void initialize(Configuration conf, InputSplit split) throws IOException, InterruptedException {
```
Shouldn't we allow data source configuration from the strings/conf here.

dalsh replied:
To deal with that I've added HikariCP as dependency for their DriverDataSource, is that ok ?

agibsonccc commented on datavec-jdbc/src/main/java/org/datavec/api/records/reader/impl/jdbc/JDBCRecordReader.java:
```diff
@@ -0,0 +1,145 @@
+                    Writable writable = JdbcWritableConverter.convert(columnValue, meta.getColumnType(i + 1));
+                    ret.add(writable);
+                } catch (SQLException e) {
+                    closeJdbc();
```
try/with resource clauses here instead?

dalsh replied:
Wouldn't that close at the end of the block in any case ? I think it may be better to leave the connection open unless something goes wrong, and let the user close it when he's finished reading .. ?

agibsonccc commented on datavec-jdbc/src/main/java/org/datavec/api/records/reader/impl/jdbc/JDBCRecordReader.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    @Override
+    public List<Writable> record(URI uri, DataInputStream dataInputStream) throws IOException {
```
throw an UnsupportedOperationExecption instead of returning null

agibsonccc commented on datavec-jdbc/src/main/java/org/datavec/api/records/reader/impl/jdbc/JDBCRecordReader.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    @Override
+    public Record nextRecord() {
```
Record shouldn't return null here. Please add the proper return and assoiated test there.

agibsonccc commented on datavec-jdbc/src/main/java/org/datavec/api/records/reader/impl/jdbc/JDBCRecordReader.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    @Override
+    public Record loadFromMetaData(RecordMetaData recordMetaData) throws IOException {
```
RecordMetaData and jdbc schemas should map. I feel like there should be a conversion here.

agibsonccc commented on datavec-jdbc/src/main/java/org/datavec/api/records/reader/impl/jdbc/JDBCRecordReader.java:
```diff
@@ -0,0 +1,145 @@
+
+    @Override
+    public List<Record> loadFromMetaData(List<RecordMetaData> recordMetaDatas) throws IOException {
+        return null;
```
Same.

agibsonccc commented on datavec-jdbc/src/main/java/org/datavec/api/records/reader/impl/jdbc/JDBCRecordReader.java:
```diff
@@ -0,0 +1,145 @@
+
+    @Override
+    public void setConf(Configuration conf) {
+
```
setConf/ and conf() should return something.

dalsh commented:
I pushed some progress, I'm going to look into InputSplits now

AlexDBlack commented on datavec-jdbc/src/main/java/org/datavec/api/util/jdbc/JdbcWritableConverter.java:
```diff
@@ -0,0 +1,57 @@
+            case Types.INTEGER:
+            case Types.SMALLINT:
+            case Types.TINYINT:
+            case Types.BIGINT:
```
BIGINT is equivalent to long I believe? Probably want to return a LongWritable.

AlexDBlack commented on datavec-jdbc/src/main/java/org/datavec/api/util/jdbc/JdbcWritableConverter.java:
```diff
@@ -0,0 +1,57 @@
+            case Types.FLOAT:
+            case Types.NUMERIC:
+            case Types.REAL:
+                return new FloatWritable((float) columnValue);
```
Are we likely to run into overflow issues here? I think float/real should be ok, but decimal/numerical could in principle overflow float min/max?

dalsh replied:
I had a similar thought recently .. Wouldn't we need something like a BigDecimalWritable for this case ?

AlexDBlack replied:
DoubleWritable would be a step up from FloatWritable, and should be fine for 99.9% of users. I'm hesitant to add another writable just for that possibility, at this point.

dalsh commented:
I've done the requested changes, added documentation, some fixes after self review and rebased on master. This is starting to look good to me too so I removed the WIP tag.

huitseeker commented on datavec-jdbc/src/test/java/org/datavec/api/records/reader/impl/TestDb.java:
```diff
@@ -0,0 +1,206 @@
+package org.datavec.api.records.reader.impl;
```
This is copy-paste from Tablesaw. Is tablesaw's test-jar usable in this context ?
https://github.com/lwhite1/tablesaw/blob/master/src/test/java/com/github/lwhite1/tablesaw/util/TestDb.java

dalsh replied:
Ah yes, this slipped my mind. I used this for convenience but I intended to replace it with more sophisticated data tables. I'll work on that and also create some tests on the data mapping which I intended to do too.

dalsh replied:
I added some unit tests which revealed some flaws in the data mapping. It's now fixed

agibsonccc commented:
Merging due to lack of activity. Thanks for the contribution!
